apps/lib/opt.c

```diff
@@ -430,6 +433,17 @@ int opt_int(const char *value, int *result)
     return 1;
 }
 
+/* Parse and return a natural number or 0; return -1 on failure. */
+int opt_nat0(void)
```

Comment:
Why `nat0` ? Why not just `nonneg` for example?
The comment is wrong about the -1 return.
The fact that zero is a valid return (and the only `opt_` parsing function where this is true) makes me say it should look like `opt_int` and fill in a `*result` and keep the 0/1 return value of everything else.

Reply:
If I enter "-5" then opt_int will parse it as -5 result will be -5, and that is what is returned.

The fact that one file, which doesn't use the standard parsing techniques that all the other files do, has an `opt_nat` is not a compelling argument to call this `opt_nat0`. It is likewise not a good justification for making this new, arguably common, code not look like what is really it's cousin `opt_int` etc.

Reply:
This is what the new function should look like:
```
int opt_nonneg(const char *value, int *result)
{
   if (!opt_int(value, result) || *result < 0) {
     print error message
     return 0;
  }
 return 1;
}
```

Reply:
Instead you have a new function, intended for all applications to use, that has a strange name and inconsistent calling sequence.

Reply:
`grep parse.*return` shows that literally *nine other functions here* work by filling in a *result and return 0 on failure, else 1.

Reply:
Consistency is nice, but for the given new function (regardless whether it's named `opt_nat0` or `opt_nonneg` or whatever) as mentioned it would just needlessly complicate its use.
The new function can only return nonnegative values on success, so returning -1 on failure is straightforward.

Not only its definition but also its use is simpler this way, e.g.,
```
if ((var = opt_nonneg()) < 0)
    goto err;
```
compared to
```
if (!opt_nonneg(opt_arg(), &var))
    goto err;
```

BTW, the latter form has the additional drawback that the type of `var` must match exactly - no implicit assignment conversion would be possible there.

Reply:
You already changed a whole bunch of lines to use your `nat0` function but changing them/others to do it and maintain consistency is too hard.  I find that ironic coming rom the person who has used "more consistent" in their PR's than anyone else :)

Implicit assignment conversion does not work unless range-checks are done, so your BTW is incorrect.

Reply:
> You already changed a whole bunch of lines to use your `nat0` function but changing them/others to do it and maintain consistency is too hard.

No, I did consistently change all uses of the former `opt_nat()`.

> I find that ironic coming rom the person who has used "more consistent" in their PR's than anyone else :)

I find consistency important for critical things, in particular to avoid semantic pitfalls,
but here convent use and simple-to-read code IMO is more important.

> Implicit assignment conversion does not work unless range-checks are done, so your BTW is incorrect.

No - for instance when `var` is of type `(unsigned) long`, there is no need for range checks.

Reply:
@richsalz,
apparently you have too much time for criticizing details of PRs of others in a project that you quit.
I find this particularly annoying because rather often the points you make are misleading or wrong 
(likely because you looked at the PRs too superficially or for other reasons miss the full picture).

I have more important things to do than answering all such stuff in detail.
So don't expect me to do so in the future :wink:

Reply:
A better way to do this would be to add a new "type" character, and handle it in `opt_next` the same way https://github.com/openssl/openssl/blob/master/apps/lib/opt.c#L802 does it.  Then you can call opt_int because you know the limits have already been verified.  (Similar to the way '/' means "must be a directory.")  It's a pity a took both 'n' and 'p'  Maybe I intended 'n' to be non-negative and just didn't finish it off.

Reply:
This is really a good thought - thanks!

So I've introduced type '`N`' for checking nonnegative option args.

And simplified in all apps:
```
            if (!opt_int(opt_arg(), &var))
                goto opthelp;
```
to
```
            var = opt_int_arg();
```
since for options of both '`p`' and '`N`' type the range has already been checked.